File: internal/protocol/datetime.go
package protocol

import (
	"regexp"
	"strconv"
)

var strictRFC3339Pattern = regexp.MustCompile(`^([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{2}):([0-9]{2}):([0-9]{2})(?:\.([0-9]{1,9}))?(Z|[+-]([0-9]{2}):([0-9]{2}))$`)

func validateRFC3339(field, value string) error {
	if err := requiredString(field, value); err != nil {
		return err
	}
	matches := strictRFC3339Pattern.FindStringSubmatch(value)
	if matches == nil {
		return invalidField(field, "must be an RFC3339 timestamp")
	}

	year := parseDatePart(matches[1])
	month := parseDatePart(matches[2])
	day := parseDatePart(matches[3])
	hour := parseDatePart(matches[4])
	minute := parseDatePart(matches[5])
	second := parseDatePart(matches[6])
	if month < 1 || month > 12 || day < 1 || day > daysInMonth(year, month) || hour > 23 || minute > 59 || second > 59 {
		return invalidField(field, "must be an RFC3339 timestamp")
	}

	if matches[8] != "Z" {
		offsetHour := parseDatePart(matches[9])
		offsetMinute := parseDatePart(matches[10])
		if offsetHour > 23 || offsetMinute > 59 {
			return invalidField(field, "must be an RFC3339 timestamp")
		}
	}
	return nil
}

func parseDatePart(value string) int {
	parsed, _ := strconv.Atoi(value)
	return parsed
}

func daysInMonth(year, month int) int {
	switch month {
	case 2:
		if isLeapYear(year) {
			return 29
		}
		return 28
	case 4, 6, 9, 11:
		return 30
	default:
		return 31
	}
}

func isLeapYear(year int) bool {
	return year%400 == 0 || (year%4 == 0 && year%100 != 0)
}

File: web/src/protocol/datetime.ts
const strictRFC3339Pattern = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,9}))?(Z|[+-](\d{2}):(\d{2}))$/

export function isRFC3339Timestamp(value: unknown): value is string {
  if (typeof value !== 'string') return false
  const matches = strictRFC3339Pattern.exec(value)
  if (matches === null) return false

  const year = BigInt(matches[1])
  const month = BigInt(matches[2])
  const day = BigInt(matches[3])
  const hour = BigInt(matches[4])
  const minute = BigInt(matches[5])
  const second = BigInt(matches[6])
  if (
    month < 1n || month > 12n ||
    day < 1n || day > daysInMonth(year, month) ||
    hour > 23n || minute > 59n || second > 59n
  ) return false

  if (matches[8] !== 'Z') {
    const offsetHour = BigInt(matches[9])
    const offsetMinute = BigInt(matches[10])
    if (offsetHour > 23n || offsetMinute > 59n) return false
  }
  return true
}

function daysInMonth(year: bigint, month: bigint): bigint {
  if (month === 2n) return isLeapYear(year) ? 29n : 28n
  if (month === 4n || month === 6n || month === 9n || month === 11n) return 30n
  return 31n
}

function isLeapYear(year: bigint): boolean {
  return year % 400n === 0n || (year % 4n === 0n && year % 100n !== 0n)
}
